File: java/egy-names/src/main/java/com/afify/egynames/engine/SearchEngine.java
package com.afify.egynames.engine;

import com.afify.egynames.index.LookupIndices;
import com.afify.egynames.model.Models;

import java.util.List;
import java.util.stream.Collectors;

public class SearchEngine {

    public static List<Models.NameInfo> search(
            Models.Gender gender,
            Models.Religion religion,
            Models.NameRole role,
            Models.FrequencyClass frequency,
            String startsWith,
            String endsWith,
            String contains,
            Double minCorpusShare,
            int maxResults,
            String sortBy,
            String dataPath) {

        List<Models.NameEntry> entries = LookupIndices.getAll(dataPath);

        boolean prefixAr = startsWith != null && LookupIndices.isArabic(startsWith);
        boolean suffixAr = endsWith != null && LookupIndices.isArabic(endsWith);
        boolean containsAr = contains != null && LookupIndices.isArabic(contains);

        List<Models.NameEntry> filtered = entries.stream().filter(e -> {
            if (gender != null && e.gender != gender && e.gender != Models.Gender.NEUTRAL) return false;
            if (religion != null && e.religion != religion && e.religion != Models.Religion.NEUTRAL) return false;
            if (role != null && e.role != role) return false;
            if (frequency != null && e.frequency != frequency) return false;
            if (minCorpusShare != null && e.corpusShare < minCorpusShare) return false;

            if (startsWith != null && !startsWith.isEmpty()) {
                if (prefixAr) {
                    if (!LookupIndices.normalizeAr(e.ar).startsWith(LookupIndices.normalizeAr(startsWith))) return false;
                } else {
                    if (!LookupIndices.normalizeEn(e.en).startsWith(LookupIndices.normalizeEn(startsWith))) return false;
                }
            }

            if (endsWith != null && !endsWith.isEmpty()) {
                if (suffixAr) {
                    if (!LookupIndices.normalizeAr(e.ar).endsWith(LookupIndices.normalizeAr(endsWith))) return false;
                } else {
                    if (!LookupIndices.normalizeEn(e.en).endsWith(LookupIndices.normalizeEn(endsWith))) return false;
                }
            }

            if (contains != null && !contains.isEmpty()) {
                if (containsAr) {
                    if (!LookupIndices.normalizeAr(e.ar).contains(LookupIndices.normalizeAr(contains))) return false;
                } else {
                    if (!LookupIndices.normalizeEn(e.en).contains(LookupIndices.normalizeEn(contains))) return false;
                }
            }

            return true;
        }).collect(Collectors.toList());

        if ("alphabetical".equalsIgnoreCase(sortBy)) {
            filtered.sort((a, b) -> a.ar.compareTo(b.ar));
        } else {
            filtered.sort((a, b) -> Double.compare(b.corpusShare, a.corpusShare));
        }

        return filtered.stream()
                .limit(maxResults)
                .map(Models.NameInfo::fromEntry)
                .collect(Collectors.toList());
    }
}

File: csharp/EgyNames/SearchEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace EgyptianNames
{
    public static class SearchEngine
    {
        public static List<NameInfo> Search(
            Gender? gender = null,
            Religion? religion = null,
            NameRole? role = null,
            FrequencyClass? frequency = null,
            string? startsWith = null,
            string? endsWith = null,
            string? contains = null,
            double? minCorpusShare = null,
            int maxResults = 50,
            string sortBy = "corpus_share",
            string? dataPath = null)
        {
            var entries = LookupIndices.GetAll(dataPath);

            bool prefixAr = startsWith != null && LookupIndices.IsArabic(startsWith);
            bool suffixAr = endsWith != null && LookupIndices.IsArabic(endsWith);
            bool containsAr = contains != null && LookupIndices.IsArabic(contains);

            var filtered = entries.Where(e =>
            {
                if (gender.HasValue && e.Gender != gender.Value && e.Gender != Gender.Neutral) return false;
                if (religion.HasValue && e.Religion != religion.Value && e.Religion != Religion.Neutral) return false;
                if (role.HasValue && e.Role != role.Value) return false;
                if (frequency.HasValue && e.Frequency != frequency.Value) return false;
                if (minCorpusShare.HasValue && e.CorpusShare < minCorpusShare.Value) return false;

                if (!string.IsNullOrEmpty(startsWith))
                {
                    if (prefixAr)
                    {
                        if (!LookupIndices.NormalizeAr(e.Ar).StartsWith(LookupIndices.NormalizeAr(startsWith))) return false;
                    }
                    else
                    {
                        if (!LookupIndices.NormalizeEn(e.En).StartsWith(LookupIndices.NormalizeEn(startsWith))) return false;
                    }
                }

                if (!string.IsNullOrEmpty(endsWith))
                {
                    if (suffixAr)
                    {
                        if (!LookupIndices.NormalizeAr(e.Ar).EndsWith(LookupIndices.NormalizeAr(endsWith))) return false;
                    }
                    else
                    {
                        if (!LookupIndices.NormalizeEn(e.En).EndsWith(LookupIndices.NormalizeEn(endsWith))) return false;
                    }
                }

                if (!string.IsNullOrEmpty(contains))
                {
                    if (containsAr)
                    {
                        if (!LookupIndices.NormalizeAr(e.Ar).Contains(LookupIndices.NormalizeAr(contains))) return false;
                    }
                    else
                    {
                        if (!LookupIndices.NormalizeEn(e.En).Contains(LookupIndices.NormalizeEn(contains))) return false;
                    }
                }

                return true;
            }).ToList();

            if (sortBy.Equals("alphabetical", StringComparison.OrdinalIgnoreCase))
            {
                filtered = filtered.OrderBy(e => e.Ar, StringComparer.Ordinal).ToList();
            }
            else
            {
                filtered = filtered.OrderByDescending(e => e.CorpusShare).ToList();
            }

            return filtered.Take(maxResults).Select(NameInfo.FromEntry).ToList();
        }
    }
}
